File: database/database-annotation-processor/src/main/java/io/koraframework/database/annotation/processor/EntityUtils.java
package io.koraframework.database.annotation.processor;

import io.koraframework.annotation.processor.common.AnnotationUtils;
import io.koraframework.annotation.processor.common.CommonUtils;
import io.koraframework.annotation.processor.common.ProcessingError;
import io.koraframework.annotation.processor.common.ProcessingErrorException;
import org.jspecify.annotations.Nullable;

import javax.lang.model.element.*;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class EntityUtils {

    public final static CommonUtils.NameConverter SNAKE_CASE_NAME_CONVERTER = originalName -> {
        var splitted = originalName.split("(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|( +)");
        return Stream.of(splitted)
            .map(String::toLowerCase)
            .collect(Collectors.joining("_"));
    };

    public static String parseColumnName(VariableElement element, CommonUtils.@Nullable NameConverter columnsNameConverter) {
        var column = AnnotationUtils.findAnnotation(element, DbUtils.COLUMN_ANNOTATION);
        var fieldName = element.getSimpleName().toString();
        if (column != null) {
            return AnnotationUtils.parseAnnotationValueWithoutDefault(column, "value");
        }
        return columnsNameConverter == null
            ? SNAKE_CASE_NAME_CONVERTER.convert(fieldName)
            : columnsNameConverter.convert(fieldName);
    }

    public static ExecutableElement findEntityConstructor(TypeElement type) throws ProcessingErrorException {
        var constructors = type.getEnclosedElements()
            .stream()
            .filter(e -> e.getKind() == ElementKind.CONSTRUCTOR)
            .map(ExecutableElement.class::cast)
            .filter(e -> e.getModifiers().contains(Modifier.PUBLIC))
            .toList();
        if (constructors.isEmpty()) {
            throw new ProcessingErrorException(List.of(new ProcessingError(entityConstructorError(type, "No public constructor was found.", "Add a public constructor that accepts all required entity fields."), type)));
        }
        if (constructors.size() == 1) {
            return constructors.get(0);
        }
        var entityConstructors = constructors.stream()
            .filter(c -> AnnotationUtils.findAnnotation(c, DbUtils.ENTITY_CONSTRUCTOR_ANNOTATION) != null)
            .toList();
        if (entityConstructors.isEmpty()) {
            throw new ProcessingErrorException(List.of(new ProcessingError(entityConstructorError(type, "More than one public constructor exists and none is marked with @EntityConstructor.", "Mark exactly one public constructor with @EntityConstructor."), type)));
        }
        if (entityConstructors.size() != 1) {
            throw new ProcessingErrorException(List.of(new ProcessingError(entityConstructorError(type, "More than one public constructor is marked with @EntityConstructor.", "Keep @EntityConstructor on exactly one public constructor."), type)));
        }
        return entityConstructors.get(0);
    }

    private static String entityConstructorError(TypeElement type, String problem, String fix) {
        return """
            Database entity constructor is invalid:
              %s

            Problem:
              %s

            Hint:
              Entity mapper generation needs exactly one constructor to create the entity from database columns.

            Fix:
              %s
            """.formatted(type.getQualifiedName(), problem, fix);
    }

}

File: database/database-symbol-processor/src/main/kotlin/io/koraframework/database/symbol/processor/EntityUtils.kt
package io.koraframework.database.symbol.processor

import com.google.devtools.ksp.symbol.KSPropertyDeclaration
import com.google.devtools.ksp.symbol.KSValueParameter
import io.koraframework.ksp.common.AnnotationUtils.findAnnotation
import io.koraframework.ksp.common.KspCommonUtils
import io.koraframework.ksp.common.parseAnnotationValue


private val SNAKE_CASE_BOUNDARY = Regex("(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|( +)")

val snakeCaseNameConverter = KspCommonUtils.NameConverter { originalName ->
    originalName.split(SNAKE_CASE_BOUNDARY)
        .map { it.lowercase() }
        .joinToString("_")
}

fun parseColumnName(valueParameter: KSValueParameter, propertyParameter: KSPropertyDeclaration, columnsNameConverter: KspCommonUtils.NameConverter?): String {
    val column = valueParameter.findAnnotation(DbUtils.columnAnnotation)
        ?: propertyParameter.findAnnotation(DbUtils.columnAnnotation)
    if (column != null) {
        return parseAnnotationValue<String>(column, "value")!!
    }

    val fieldName = valueParameter.name!!.asString()
    if (columnsNameConverter != null) {
        return columnsNameConverter.convert(fieldName)
    }

    return snakeCaseNameConverter.convert(fieldName)
}
